### PR TITLE
Add user registration

### PR DESCRIPTION
**Pull request by KostyaKobel**
## Checklist:

- [x] Added specs for all new ruby code
- [x] Green rspec run
- [x] Green rubocop
- [x] PR review from teammate

# What?
This is the Authorization page and Registration page User.

## Why?
This is necessary so that the user can go to his profile page and use our site for their respective purposes.

## How?
Install gem devise and configure him with user controller invoked on root (/) requests. A Slim view for the autorisation and registration in app/views/devise/registration/new.html.slim and app/views/devise/sessions/new.html.slim And some stylesheets for the pages as well.

## Testing?
Yes I wrote tests for the User model.

## Screenshots [(optional)](url)
http://localhost:3000/users/sign_in
SignUp2
SignUp3
SignUp4
SignUp5
SignUp10
Forgott_Password
Resend_confirmation

### CODE REVIEW

**Review by Hunter2065 — CHANGES_REQUESTED** [review]

**Review by Hunter2065 — CHANGES_REQUESTED** [review]
Will be fine if you unmerge 'demo1'

File: config/environments/development.rb [hunk 1]
```diff
@@ -70,6 +70,14 @@
   # Use an evented file watcher to asynchronously detect changes in source code,
   # routes, locales, etc. This feature depends on the listen gem.
   config.file_watcher = ActiveSupport::EventedFileUpdateChecker
+  config.action_mailer.default_url_options = { :host => 'localhost:3000' }
+
+  config.action_mailer.delivery_method = :smtp
+  host = 'localhost:3000'
+  config.action_mailer.default_url_options = { :host => 'localhost:3000', protocol: 'http' }
```
Comment (andrjuha01) [inline config/environments/development.rb:77]: you have it duplicated here
https://github.com/montashevych/Hostmeplease/pull/67/files#diff-d3c4b3f41072daa416f1920511e9b2e26caea8c5cec0a14cb9508589a4dafa47R73

File: config/environments/development.rb [hunk 2]
```diff
@@ -70,6 +70,14 @@
   # Use an evented file watcher to asynchronously detect changes in source code,
   # routes, locales, etc. This feature depends on the listen gem.
   config.file_watcher = ActiveSupport::EventedFileUpdateChecker
+  config.action_mailer.default_url_options = { :host => 'localhost:3000' }
+
+  config.action_mailer.delivery_method = :smtp
+  host = 'localhost:3000'
```
Comment (andrjuha01) [inline config/environments/development.rb:76]: remove this one

File: config/environments/production.rb [hunk 3]
```diff
@@ -68,6 +68,22 @@
   # Set this to true and configure the email server for immediate delivery to raise delivery errors.
   # config.action_mailer.raise_delivery_errors = false
 
+  config.action_mailer.delivery_method = :smtp
+  host = 'https://hostmeplease.herokuapp.com/' #replace with your own url
+  config.action_mailer.default_url_options = { host: host }
+  heroku config:set GOOGLE_LOGIN="aleksanderis2121@gmail.com"
```
Comment (andrjuha01) [inline config/environments/production.rb:74]: nooo,
use console to do that

File: config/environments/production.rb [hunk 4]
```diff
@@ -68,6 +68,22 @@
   # Set this to true and configure the email server for immediate delivery to raise delivery errors.
   # config.action_mailer.raise_delivery_errors = false
 
+  config.action_mailer.delivery_method = :smtp
+  host = 'https://hostmeplease.herokuapp.com/' #replace with your own url
```
Comment (andrjuha01) [inline config/environments/production.rb:72]: remove this variable and set host here: https://github.com/montashevych/Hostmeplease/pull/67/files#diff-da60b4e96eff2b132991226d308949e23f4ef3aad45ad59edd09cbc32cc6251eR73

File: spec/mailers/previews/user_mailer_preview.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# Preview all emails at http://localhost:3000/rails/mailers/user_mailer
+class UserMailerPreview < ActionMailer::Preview
+  def new_user_email
+    # Set up a temporary user for the preview
+    user = User.new(FirstName: 'Ivan', LastName: 'Bogun', email: 'ivan@gmail.com', message: 'I want to log in!')
```
Comment (andrjuha01) [inline spec/mailers/previews/user_mailer_preview.rb:5]: do we need temporary user here

File: spec/models/user_spec.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      expect(valid_user).to be_valid
+    end
+
+    it 'creates an user with owner role' do
+      expect(valids_user).to be_valid
+    end
+  end
+
+  context 'when not valid' do
+    let(:unvalid1_user) { build(:user, first_name: '    ') }
```
Comment (andrjuha01) [inline spec/models/user_spec.rb:18]: unvalid -> invalid
Prefer
user_without_name
user_without_password 
Than just invalid user

File: db/migrate/20210527133127_devise_create_users.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      # t.datetime :current_sign_in_at
+      # t.datetime :last_sign_in_at
+      # t.string   :current_sign_in_ip
+      # t.string   :last_sign_in_ip
+
+      ## Confirmable
+      t.string   :confirmation_token
+      t.datetime :confirmed_at
+      t.datetime :confirmation_sent_at
+      t.string   :unconfirmed_email # Only if using reconfirmable
```
Comment (andrjuha01) [inline db/migrate/20210527133127_devise_create_users.rb:31]: are we using reconfirmable?
If no - remove it

File: db/migrate/20210527133127_devise_create_users.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
+class DeviseCreateUsers < ActiveRecord::Migration[6.1]
+  def change
+    create_table :users do |t|
+      ## Database authenticatable
+      t.string :first_name,             null: false, default: ""
```
Comment (andrjuha01) [inline db/migrate/20210527133127_devise_create_users.rb:5]: Can user be created without first name?

File: db/migrate/20210527133127_devise_create_users.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+      t.string :last_name,              null: false, default: ""
+      t.string :role,                   null: false, default: "consumer"
+      t.string :email,                  null: false, default: ""
```
Comment (andrjuha01) [inline db/migrate/20210527133127_devise_create_users.rb:8]: can user be created without email?

**Review by andrjuha01 — CHANGES_REQUESTED** [review]

File: spec/models/user_spec.rb [hunk 10]
```diff
@@ -2,43 +2,43 @@
...
     end
   end
 
   context 'when not valid' do
-    let(:unvalid1_user) { build(:user, first_name: '    ') }
-    let(:unvalid2_user) { build(:user, first_name: '1' * 100) }
-    let(:unvalid3_user) { build(:user, email: '') }
-    let(:unvalid4_user) { build(:user, password: '') }
-    let(:unvalid5_user) { build(:user, role: '') }
+    let(:user_without_name_empty) { build(:user, first_name: '    ') }
```
Comment (andrjuha01) [inline spec/models/user_spec.rb:18]: user_with_empty_name.
Check if wording make sense

File: spec/factories/users.rb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :user do
+    first_name  { Faker::Name.first_name }
+    password    { Faker::Internet.password(min_length: 6) }
+    email       { Faker::Internet.unique.email }
+  end
+end
```
Comment (VanyaNovak) [inline spec/factories/users.rb:7]: You can delete this file because it does not belong to the user model. on my branch it is already changed

**Comment by VanyaNovak** [comment]
Add titles for pages

File: app/views/devise/passwords/new.html.slim [hunk 12]
```diff
@@ -1,3 +1,5 @@
+- provide :title, "Forgott Password"
```
Comment (VanyaNovak) [inline app/views/devise/passwords/new.html.slim:1]: Change "Forgott" to "Forgot"

File: app/views/devise/passwords/edit.html.slim [hunk 13]
```diff
@@ -1,3 +1,5 @@
+- provide :title, "Forgott Password"
```
Comment (VanyaNovak) [inline app/views/devise/passwords/edit.html.slim:1]: And here change "Forgott" to "Forgot"

**Review by VanyaNovak — APPROVED** [review]
LGTM!

File: db/migrate/20210527133127_devise_create_users.rb (hunk 7 above)
Comment (KostyaKobel) [reply db/migrate/20210527133127_devise_create_users.rb:31]: Yes we will be use it

**Review by Hunter2065 — APPROVED** [review]
okay, but I do not vouch for it

File: .github/workflows/workflow.yml [hunk 14]
```diff
@@ -46,8 +46,11 @@ jobs:
         run: sudo apt-get install libpq-dev
       - name: Install dependencies
         run: |
-          gem install bundler
+          gem install bundler
```
Comment (andrjuha01) [inline .github/workflows/workflow.yml:49]: remove trailing spaces

File: config/environments/production.rb [hunk 15]
```diff
@@ -68,6 +68,20 @@
...
 
+  config.action_mailer.delivery_method = :smtp
+  
+  config.action_mailer.default_url_options = { host: ENV['HEROK_HOST'] }
+    
+  # SMTP settings for gmail
+  config.action_mailer.smtp_settings = {
+    :address              => "smtp.gmail.com",
+    :port                 => 587,
+    :user_name            => ENV['GMAIL_LOGIN'],
```
Comment (andrjuha01) [inline config/environments/production.rb:79]: nice

File: config/environments/production.rb [hunk 16]
```diff
@@ -68,6 +68,20 @@
   # Set this to true and configure the email server for immediate delivery to raise delivery errors.
   # config.action_mailer.raise_delivery_errors = false
 
+  config.action_mailer.delivery_method = :smtp
+  
+  config.action_mailer.default_url_options = { host: ENV['HEROK_HOST'] }
```
Comment (andrjuha01) [inline config/environments/production.rb:73]: HEROKU_HOST

File: docker-compose.yml [hunk 17]
```diff
@@ -3,7 +3,7 @@ services:
   db:
     image: postgres
     ports:
-      - "5432:5432"
+      - "5431:5431"
```
Comment (andrjuha01) [inline docker-compose.yml:6]: hmm, that's something wrong here.
Inner port should be 5432, outer can be anything.

File: app/assets/stylesheets/fonts.scss [hunk 18]
```diff
@@ -1,2 +1,2 @@
 @import url('https://fonts.googleapis.com/css2?family=Montserrat:wght@100;700&display=swap');
-@import url('https://fonts.googleapis.com/css2?family=Material+Icons');
+@import url('https://fonts.googleapis.com/css2?family=Material+Icons');
```
Comment (andrjuha01) [inline app/assets/stylesheets/fonts.scss:2]: always leave last line empty
https://stackoverflow.com/a/2287990

File: app/views/layouts/application.html.slim [hunk 19]
```diff
@@ -12,6 +12,10 @@ html
     = stylesheet_link_tag 'https://fonts.googleapis.com/css2?family=Material+Icons', media: 'all', 'data-turbolinks-track': 'reload'
     = javascript_pack_tag 'application', 'data-turbolinks-track': 'reload'
     = javascript_pack_tag 'site', 'data-turbolinks-track': 'reload'
+    /! Development version
+    script src="https://unpkg.com/@popperjs/core@2/dist/umd/popper.js" 
+    /! Production version
+    script src="https://unpkg.com/@popperjs/core@2"
```
Comment (andrjuha01) [inline app/views/layouts/application.html.slim:18]: leave only one

File: Gemfile [hunk 20]
```diff
@@ -33,14 +31,16 @@ gem 'slim', '~> 4.1'
...
 gem 'will_paginate', '~> 3.3.0'
 
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
-  gem 'factory_bot', '~> 6.2'
+  gem 'factory_bot_rails', '~> 6.2'
   gem 'faker', '~> 2.18'
   gem 'rspec-rails', '~> 5.0.0'
+  gem 'sqlite3', '~> 1.4'
```
Comment (andrjuha01) [inline Gemfile:43]: remove

File: app/assets/stylesheets/alert.scss [hunk 21]
```diff
@@ -0,0 +1,12 @@
+.alert {
+  top: 16px;
```
Comment (andrjuha01) [inline app/assets/stylesheets/alert.scss:2]: remove app/assets/images/place-background.png

File: app/controllers/users_controller.rb [hunk 22]
```diff
@@ -0,0 +1,20 @@
+class UsersController < ApplicationController
+  before_action :authenticate_user!
+  def index
```
Comment (andrjuha01) [inline app/controllers/users_controller.rb:3]: add new line

File: app/views/layouts/application.html.slim [hunk 23]
```diff
@@ -8,7 +8,25 @@ html
     = csrf_meta_tags
     = csp_meta_tag
     = stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload'
+    = stylesheet_link_tag 'https://fonts.googleapis.com/css2?family=Montserrat:wght@100;700&display=swap', media: 'all', 'data-turbokinks-track': 'reload'
```
Comment (andrjuha01) [inline app/views/layouts/application.html.slim:11]: move to separate file

File: config/application.rb [hunk 24]
```diff
@@ -1,6 +1,7 @@
 require_relative 'boot'
```
Comment (andrjuha01) [inline config/application.rb:1]: app/views/users/index.html.slim
app/views/users/show.html.slim
are they empty?

File: config/routes.rb [hunk 25]
```diff
@@ -1,6 +1,9 @@
 Rails.application.routes.draw do
   resources :places
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
-
   root 'home#index'
+
+  resources :places
```
Comment (andrjuha01) [inline config/routes.rb:6]: remove from line #2 same definition

File: db/migrate/20210527133127_devise_create_users.rb [hunk 26]
```diff
@@ -0,0 +1,49 @@
+class DeviseCreateUsers < ActiveRecord::Migration[6.1]
+  def change
+    create_table :users do |t|
+      ## Database authenticatable
+      t.string :first_name
+      t.string :last_name              
+      t.string :role,                   null: false, default: "consumer"
+      t.string :email                  
+      t.string :encrypted_password,     null: false, default: ""
```
Comment (andrjuha01) [inline db/migrate/20210527133127_devise_create_users.rb:9]: can user be created without password

File: package-lock.json [hunk 27]
```diff
@@ -0,0 +1,8930 @@
+{
```
Comment (andrjuha01) [inline package-lock.json:1]: do we need this file?

Comment (KostyaKobel) [reply package-lock.json:1]: Yes. I had an error not booting normally bootstrap and css styles per page

File: config/application.rb (hunk 24 above)
Comment (KostyaKobel) [reply config/application.rb:1]: Yes, we don't have a user profile yet. That's why they're empty now.

**Review by andrjuha01 — APPROVED** [review]

File: spec/rails_helper.rb [hunk 28]
```diff
@@ -5,6 +5,7 @@
 # Prevent database truncation if the environment is production
 abort('The Rails environment is running in production mode!') if Rails.env.production?
 require 'rspec/rails'
+require 'support/factory_bot'
```
Comment (andrjuha01) [inline spec/rails_helper.rb:8]: is this required?

**Review by andrjuha01 — APPROVED** [review]

File: spec/rails_helper.rb (hunk 28 above)
Comment (KostyaKobel) [reply spec/rails_helper.rb:8]: Fixed

**Pull request merged**